File: #0703.kth-largest-element-in-a-stream.cpp
class KthLargest {
public:
    KthLargest(int k, vector<int>& nums) {

        kthLargest = vector<int>();
        if (nums.size() > 0){
            std::sort(nums.begin(), nums.end());
            for (vector<int>::reverse_iterator rit = nums.rbegin();
                 rit < nums.rbegin() + k && rit != nums.rend(); rit++) {
                kthLargest.push_back(*rit);
            }
        }
        for (int i = kthLargest.size(); i < k; i++) {
            kthLargest.push_back(INT_MIN);
        }
    }
    
    int add(int val) {

        int k = kthLargest.size();
        kthLargest.push_back(val);
        for (int i = 0; i < k; i++) {
            if (val >= kthLargest[i]) {
                kthLargest.insert(kthLargest.begin() + i, val);
                break;
            }
        }
        kthLargest.resize(k);
        return kthLargest[k - 1];
    }

private:
    vector<int> kthLargest;
};

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */

File: #0703.kth-largest-element-in-a-stream.js
/**
 * @param {number} k
 * @param {number[]} nums
 */
var KthLargest = function(k, nums) {
    this.main = new MinPriorityQueue();
    this.size = k; 
    for (let num of nums) {
        this.add(num);
    }
};

/** 
 * @param {number} val
 * @return {number}
 */
KthLargest.prototype.add = function(val) {
    this.main.enqueue(val);
    if (this.main.size() > this.size) {
        this.main.dequeue();
    }
    return this.main.front().element;
};

/** 
 * Your KthLargest object will be instantiated and called as such:
 * var obj = new KthLargest(k, nums)
 * var param_1 = obj.add(val)
 */
